zaza/openstack/charm_tests/nova/tests.py

```diff
@@ -141,6 +142,107 @@ def test_930_check_virsh_default_network(self):
+        nova_sentries = zaza.model.get_units(
+            'nova-compute', model_name=self.model_name)
+        for nova_sentry in nova_sentries:
+            nova_unit_name = nova_sentry.entity_id
+
+            marked_done = False
+            for ceph_sentry in ceph_sentries:
+                ceph_unit_name = ceph_sentry.entity_id
+                logging.info('Checking ceph broker messaging between '
+                             'units {} and {}'.format(nova_unit_name,
+                                                      ceph_unit_name))
+                output = zaza_juju.get_relation_from_unit(
+                    nova_unit_name, ceph_unit_name, 'client')
+                logging.info('Output of relation is: {}'.format(output))
+                broker_key = 'broker-rsp-{}'.format(
+                    nova_unit_name.replace('/', '-'))
+                if broker_key in output.keys():
+                    broker_rsp_string = output[broker_key]
+                    broker_rsp_dict = json.loads(broker_rsp_string)
+                    unit_number = nova_unit_name.partition('/')[2]
+                    key = "unit_{}_ceph_broker_action." \
+                          "nova_compute_restart".format(unit_number)
+                    command = 'cd hooks;python -c "' \
+                              'from charmhelpers.core.unitdata import kv;' \
+                              'kvstore = kv();' \
+                              'val = kvstore.get(key=\\"{}\\");' \
+                              'print(val)"'.format(key)
+                    output = zaza.model.run_on_unit(
+                        nova_unit_name, command).get('Stdout').strip()
+                    if broker_rsp_dict['request-id'] in output:
+                        logging.info(
+                            'Found request-id in output: '
+                            '{}'.format(broker_rsp_dict['request-id']))
+                        marked_done = True
+
+            if not marked_done:
+                msg = "Ceph broker message was not processed " \
+                      "successfully for {}".format(nova_unit_name)
+                raise Exception("Test failed: {}".format(msg))
+
+        # Make sure pool was created
+        ceph_sentry = ceph_sentries[0]
+        pools = zaza.model.run_on_unit(
+            ceph_sentry.entity_id,
+            'sudo ceph osd lspools').get('Stdout').strip()
+        logging.info('Pools found: {}'.format(pools))
+        if "nova" not in pools:
+            msg = "Nova ceph pool was not created successfully"
+            raise Exception("Test failed: {}".format(msg))
+
+        logging.info('test_1000_ceph_broker_message_marked_done successful')
+
+    def test_1001_config_changed_cause_restarts(self):
+        """Verify service is not restarted on config-changed hook."""
+        self._test_hook_cause_restarts('config-changed')
+        logging.info('test_1001_config_changed_cause_restarts successful')
```

Comment:
I'm a bit dubious about this test; it's forcing a config-changed hook run by calling the actual hook, rather than doing some config change and recording whether the proposed thing change.

I think the test is probably fragile.  e.g. if nova-compute were optimised such that no actual config change caused a restart regardless of whether the actual hook is called, then this test would fail -- it's assuming implementation rather than behaviour.

There is a config_changed context manager (`config_change()`) which can be used, or even the `restart_on_changed_debug_oslo_config_file()` helper which can check for restarts.  Please review the `charm_tests/test_utils.py` for a suitable helper?

Reply:
As I stated in my comment, I believe this test is valuable. Please check my comment with extra context and let me know what you think.